### PR TITLE
BUG: Fix downcast argument for DataFrame.fillna()

### LINKED ISSUE #15277: DataFrame.fillna() downcast argument does not work when the value for fillna is a dict instead of a scalar
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

df = pd.DataFrame({'a': [1, None]})

df1 = df.fillna(0, downcast='infer')
df2 = df.fillna({'a': 0}, downcast='infer')

print(df1)
print(df2)
```
#### Problem description

When using the method `DataFrame.fillna()`, the `downcast` option is not working when the `value` for `fillna` is a dict instead of a scalar.

#### Expected Output

In both previous DataFrame examples, the column should be downcasted to integer dtype instead of float.

#### Output of ``pd.show_versions()``

<details>
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-107-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: ca_ES.UTF-8

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.2
setuptools: 27.2.0.post20161106
Cython: 0.24.1
numpy: 1.11.1
scipy: 0.18.1
statsmodels: 0.6.1
xarray: None
IPython: 5.1.0
sphinx: 1.4.6
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.6.1
blosc: None
bottleneck: 1.1.0
tables: 3.2.3.1
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: 2.3.2
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.3
lxml: 3.6.4
bs4: 4.5.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.13
pymysql: None
psycopg2: None
jinja2: 2.8
boto: 2.42.0
pandas_datareader: 0.2.1
</details>

### PR DESCRIPTION
**Pull request by albertvillanova**
- [X] closes #15277 
 - [X] tests added / passed
 - [x] passes git diff upstream/master | flake8 --diff
 - [X] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
needs tests, pls add for both series & dataframe (in pandas/tests/series/test_missing.py and similar for frame).

add alongside the existing tests

pls add a whatsnew note (bug fixes for 0.20.0)

File: pandas/tests/series/test_missing.py [hunk 1]
```diff
@@ -268,6 +268,19 @@ def test_datetime64tz_fillna_round_issue(self):
 
         assert_series_equal(filled, expected)
 
+    def test_fillna_downcast(self):
+        # infer int64 from float64
+	s = pd.Series([1., np.nan])
```
Comment (jreback) [inline pandas/tests/series/test_missing.py]: this is a flake error :<

**Comment by jreback** [comment]
lgtm. though you are going to fail linting because of indentation :< ping on green

**Comment by albertvillanova** [comment]
Is it OK now or some additional changes are needed? Should I squash all the commits into a single one?

**Comment by jreback** [comment]
thanks @albertvillanova 

squashing is really no longer necessary as our merging process will squash (and retain the commit links).

**Pull request closed without merging**